libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/arrow/api/MapboxRouteArrowApi.kt

```diff
@@ -61,101 +70,194 @@ import com.mapbox.turf.TurfMeasurement
+                                .arrowHeadFeatureCollection
+                                .features()
+                                ?.firstOrNull()
+                        )
+                    )
+                }
+            }
+        }
+
+    /**
+     * Adds an arrow to the map made up of the points submitted. An arrow is made up of at least
+     * two points. The direction of the arrow head is determined by calculating the bearing
+     * between the last two points submitted. Each call will add a new arrow.
+     *
+     * @param points the points that make up the arrow to be drawn
+     *
+     * @return a Expected<ArrowAddedValue, RouteArrowUpdateError>
+     */
+    fun addArrow(points: List<Point>): Expected<ArrowAddedValue, RouteArrowUpdateError> {
+        if (points.size < RouteConstants.TWO_POINTS) {
+            return Expected.Failure(
+                RouteArrowUpdateError.InvalidPointError(
+                    "An arrow must have at least 2 points.",
+                    null
+                )
             )
-        val invalidCurrentStepPoints = routeProgress.currentLegProgress == null ||
-            routeProgress.currentLegProgress!!.currentStepProgress == null ||
-            routeProgress.currentLegProgress!!.currentStepProgress!!.stepPoints == null ||
-            routeProgress.currentLegProgress!!.currentStepProgress!!.stepPoints!!.size <
-            RouteConstants.TWO_POINTS
-
-        val visibilityChanges = if (invalidUpcomingStepPoints || invalidCurrentStepPoints) {
-            getHideArrowModifications()
-        } else {
-            getShowArrowModifications()
         }
 
-        val newManeuverPoints = if (!invalidUpcomingStepPoints && !invalidCurrentStepPoints) {
-            RouteArrowUtils.obtainArrowPointsFrom(routeProgress)
+        return if (arrows.flatten().intersect(points).isEmpty()) {
+            arrows.add(points)
+            return Expected.Success(
+                ArrowAddedValue(
+                    getShaftFeatureCollection(),
+                    getArrowHeadFeatureCollection()
+                )
+            )
         } else {
-            listOf()
+            Expected.Failure(
+                RouteArrowUpdateError.AlreadyPresentError(
+                    "An arrow with one or more of those points already exists",
+                    null
+                )
+            )
+        }
+    }
+
+    /**
+     * Will remove any arrow having one or more points contained in the points submitted.
+     * To remove a previously added arrow it isn't necessary to submit all of the points of the
+     * previously submitted arrow(s). Instead it is necessary only to submit at least one point
+     * for each previously added arrow that should be removed.
+     *
+     * @param points one or more points used as criteria for removing arrows from the map
+     *
+     * @return a [RemoveArrowValue]
+     */
+    fun removeArrow(points: List<Point>): RemoveArrowValue {
```

Comment:
Same as above. Should we model the arrows differently? Wondering if we'd like to implement the concept of "arrow" (instead of a list of points) and remove that particular arrow - that way we'll be explicit and won't need the logic around removing one point from the list will end up removing the "arrow" (list). What was your original thinking around having a `List`?

Reply:
I thought of abstracting this to the concept of an "arrow" type in the beginning.  In then end I decided it might complicate things more for the developer.

An arrow is made up of 2 or more points with the arrow head placed at the end of the line.  If we use the concept of an Arrow type then the developer has to keep track of their arrows so that they can remove them.  That seems more complex than simply keeping track of a single point that might be part of a step that enters or exits a round about for example given that the route is already made up of points.  It seems easier to tell the API, remove any arrows containing point X.

Reply:
Per our discussion a ManeuverArrow type was added.